### ISSUE TITLE
ffill+bfill是不是还是会在测试集里引入未来信息呀？

### ISSUE THREAD

**Issue by YHY-10** [SJTU-DMTai/MASTER#25]
感谢作者的工作，很有启发性
最近follow了MASTER、MATCC等系列工作，尽管作者更新了opensouce data，但是pickle文件里里面还是使用ffill+bfill，应该还是会引入未来信息？

**Comment by LITONG99** [comment]
No. If you don't know the label, you cannot compare the model prediction to it.

**Comment by weituo2002** [comment]
会有影响，我刚做了实验的，就是将重复行完全去除之后，IC值将降到0.04左右
你只要在base_model.py中加入一个删除重复行的函数，则个重复行就是由于填充引起的：
#删除有重复行的股票数据
def drop_duplicates(x, tolerance=1e-10):
    """
    检查每只股票的8天数据中是否存在重复的特征行
    Args:
        x: 特征张量 shape (N, T, F)，其中:
           N 是股票数量
           T=8 是每只股票的交易日数
           F=221 是每天的特征数
    Returns:
        mask: 布尔掩码，标记不含重复行的股票位置
    会检查每只股票的8天数据中是否有重复的特征行
    如果发现某只股票在任意两天的特征完全相同，就将这只股票的所有数据删除
    保持了数据的完整性，要么保留一只股票的全部8天数据，要么完全删除
    """
    N, T, F = x.shape
    # 创建掩码，初始全为True
    mask = torch.ones(N, dtype=torch.bool, device=x.device)

    # 对每只股票检查其8天的数据
    for stock_idx in range(N):
        stock_data = x[stock_idx]  # shape: (8, 221)

        # 检查这只股票的8天数据中是否有重复行
        found_duplicate = False
        feature_dict = {}

        # 比较8天中任意两天是否相同
        for day_idx in range(T):
            feat_key = tuple(stock_data[day_idx].cpu().numpy())
            if feat_key in feature_dict:
                # 发现重复行，标记这只股票需要删除
                found_duplicate = True
                break
            else:
                feature_dict[feat_key] = day_idx

        if found_duplicate:
            # 如果发现重复行，将这只股票标记为False
            mask[stock_idx] = False

    return mask
    
   然后在 def train_epoch(self, data_loader):中修改每个批次的循环，增加删除重复行的函数：
    for batch_idx, data in enumerate(data_loader):
            data = torch.squeeze(data, dim=0)
            feature = data[:, :, 0:-1].to(self.device)  # shape: (N, 8, 221)
            label = data[:, -1, -1].to(self.device)  # shape: (N,)

            #print(f"Batch {batch_idx} - Original size: {feature.shape[0]}")

            # 处理极端值
            mask, label = drop_extreme(label)
            feature = feature[mask, :, :]
            #print(f"After extreme drop: {feature.shape[0]}")

            if feature.shape[0] > 0:
                # 处理重复行
                dup_mask = drop_duplicates(feature)
                #print(f"Duplicate mask sum: {dup_mask.sum()} out of {len(dup_mask)}")
                feature = feature[dup_mask, :, :]
                label = label[dup_mask]
                #print(f"After duplicate drop: {feature.shape[0]}")

                if feature.shape[0] > 0:
                    # 标准化处理
                    label = zscore(label)
                   # print(f"Label after zscore - mean: {label.mean():.4f}, std: {label.std():.4f}")
                   
   最终的结果会下降不少：
IC: 0.0474 pm 0.0061
ICIR: 0.2996 pm 0.0499
RIC: 0.0513 pm 0.0067
RICIR: 0.3015 pm 0.0533

**Comment by initsownright** [comment]
![image](https://github.com/user-attachments/assets/03c4a12f-dea6-4366-aacb-9c2bda7a82ce)

我用的自己的数据集，特征只有20+。这是超参优化过程，5个随机数种子上的平均结果。。

**Comment by LITONG99** [comment]
Thank you for the valuable experiment. It is very inspiring. 

However, it seems the experiments are not addressing the "potential future information" problem. And, I do not think the proposed preprocessing is the best, or even rational way.

For example, if I observed values on t=[0, 1, 2, 3, 4, 6, 7 ],  but missing values at t=5. Filling up t=5 with the value from t=6 will not introduce future information when we consider prediction at t=8, so why must I abandon the valuable observation? It is not surprising that the performance degrades if you drop instead of filling up, as you abandon a lot of information. As we stated in the paper, the problem is already in data shortage.

**Comment by weituo2002** [comment]
> Thank you for the valuable experiment. It is very inspiring.
> 
> However, it seems the experiments do not address the "potential future information" problem. And, I do not think the proposed preprocessing is the best or even rational way.
> 
> For the specific experiments, if I observed values on t=[0, 1, 2, 3, 4, 6 ], but missing values at t=5. Filling up t=5 with the value from t=6 will not introduce future information when we consider prediction at t=7, so why must I abandon the valuable observation? **It is not surprising that the performance degrades if you brutely drop values instead of filling up**, because you abandon a lot of information. As we discussed in the paper, the domain is in data shortage, so I do not see a clear purpose of the proposed experiments.
> 
> I have been frequently asked similar questions to 'prevent future info'. For instance, whether the code should include a **casual mask**. The key here is to consider how the training samples are constructed. The training samples are organized in a **sliding window** style instead of **auto-regression**, as designed by the qlib to suit the stock domain.
> 
> **Sliding window**: Sample 1: input t=[0, 1, 2], compute loss for t=3. Sample 2: input t=[-1, 0, 1], compute loss for t=2. Sample 3: input t=[-2, -1, 0], compute loss for t=1. Each sample is independent of the others (but in parallel during training to accelerate.)
> 
> **Auto-regression**: Sample 1: input t=[0, 1, 2], compute loss for t=1, 2, 3 (usually shift 1). Here one would apply 'causal mask' to mask out future values. Specifically, mask out values as t=1, 2 to predict for t=1; mask out values at t=2 to predict for t=2; no mask out to predict for t=3. This setting is frequently used in NLP since the original sentences are of different lengths, which makes it hard to efficiently put them into the sliding window style.
> 
> All in all, it seems to be a basic question regarding training in this domain, which is not derived from our methods.
作者可以看看我在MATCC中的实验结果，https://github.com/caozhiy/MATCC/issues/4#top  ，DFT25和MATCC都是在您这篇论文基础上改的，但是他们在预处理阶段用qlib的预处理器执行了删除极值的工作，这样就引入了更多的重复值，可以通过实验看到，在股票数据中只要重复值引入越多，那么取得的ic成绩就越好，这就是dft和matcc论文的ic值能够达到1+的原因，反过来就说明了股票数据中使用重复值是一种强烈的伪特征，只要去掉重复数据结果就大大降低，也就是说，ic的值完全可以通过增加重复数据的占比不断在测试集上提升成绩。
我在DTF-25中的详细分析评论已经被作者删除看不到了，现在只有MATCC中的部分结论评论如下：
我仔细打印了预处理阶段的数据集和训练阶段的数据集，确实因为DropExtremeLabel预处理器造成很多NaN行的存在，从而因为qlib的机制会引入很多的重复填充行，但是其中的引起数据重复填充的情况还是要分为两类：一类是因为股票数据距离开始日期不足8个交易日，占总训练集的6%；另一类是因为DropExtremeLabel标签极值删除预处理器带来的，占总训练集的15.4%。
最终做了实验，将dropextremelabel预处理器去掉之后，并且RobustZScoreNorm改成绝对不引入未来信息的，只使用到当前交易日为止的历史数据的滑动窗口的RobustZScoreNorm，做75次epoch，验证集的ic值始终在0.01到0.02之间甚至为负的情况都有，而测试集始终不会超过0.03到0.04左右，大幅低于0.1+这种论文结果，看来引入了15%的重复数据在数据集中，对于结果的影响是巨大的
详细分析如下：
1、在数据预处理阶段generate_dataset.py中
TSDataSampler调用初始化__init__，在初始化阶段
self.idx_df, self.idx_map = self.build_index(self.data)
会调用build_index:
1）idx_df = pd.Series(range(data.shape[0]), index=data.index, dtype=object)
将格式为<datetime日期，instrument股票代码,feature特征（多个）,label标签>，其中<datetime日期，instrument股票代码>是二级索引的原始数据集从0开始编号，得到每行数据的一维索引号
2）idx_df = lazy_sort_index(idx_df.unstack())
将<datetime日期，instrument股票代码>二级索引，再转变为行是日期，列是每个股票代码的矩阵：
<datetime日期，instrument股票代码1，instrument股票代码2，……，instrument股票代码N>
该矩阵的元素就是，元数组中每一行的一维索引号
对于当日没有数据的股票，在矩阵中被填入NaN值
并通过self.data_index = deepcopy(self.data.index)保存了原始数据的二级索引

2、在模型训练阶段train_model_MATCC.py
在train_epoch函数中，通过循环for data in data_loader，调用了TSDataSampler的__getitem__，其中将一维行号索引转成了索引矩阵<datetime,instrument1,……instrumentN>的行列号然后调用了_get_indices来获取每个股票记录的连续8个交易日的数据：
indices = self.idx_arr[max(row - self.step_len + 1, 0) : row + 1, col]
对数据集交易日开始不满8个交易日的股票索引都用NaN值填充成连续8个交易日：
if len(indices) < self.step_len:
indices = np.concatenate([np.full((self.step_len - len(indices),), np.nan), indices])
对于被DropExtremeLabel所删除的股票记录，在索引矩阵<datetime,instrument1,……instrumentN>中的索引值都是NaN。
然后这两种情况都会被用前后的索引进行填充，最后导致的取数据相同的索引会得到重复的行
if self.fillna_type == "ffill":
indices = np_ffill(indices)
elif self.fillna_type == "ffill+bfill":
indices = np_ffill(np_ffill(indices)[::-1])[::-1]

经过数据统计，不使用去标签极值过滤器DropExtremeLabel，有6%的股票数据因为距离开始日期不足8个交易日存在NaN值



而使用了去标签极值过滤器DropExtremeLabel，有21.4%的股票数据因为距离开始日期不足8个交易日和因为被删除了标签值在极值范围内的交易日数据而存在NaN值，其中因为被删除了标签值在极值范围内的交易日数据而存在NaN值的占比是21.4%-6%=15.4%

![Image](https://github.com/user-attachments/assets/8c74594d-a632-4cc0-98b9-3b44022b9014)
因此因为距离开始日期不足8个交易日而引入的重复数据占比是6%
而因为删除了标签值在极值范围内的交易日数据而引入的重复数据占比是15.4%
那么最终的问题就是这些总共21.4%的重复数据会对结果有非常大的影响。

**Comment by LITONG99** [comment]
There seem to be a lot of experiments, but I am not sure what the points are. And I do not think your post is related to my previous answer......May I suggest that everyone organize and shorten their posts carefully so I can fully understand the points? 

Sorry that I do not have the time to follow MATCC and discuss its details with you. Nevertheless, let me guess that you are making the claim:
	Introducing **artificial training data** (preprocessing includes data imputation, dropping samples with extreme labels, dropping NaN data, etc. ) is harmful. Although **they improve IC on the test set**, people **cannot trust the improvements** since the data are not authentic.

Here I will address my opinion based on MASTER instead of any other works. 

(1) We can introduce artificial training data as long as the testing environment is fair. You can generate 100% artificial data to train the model and report the performance as long as you do not tune your methods on the test data. To trust or not is subjective.

(2) The preprocessing of MASTER helps to improve performance, but it may not be optimal concerning improvements in the test data. We just follow the Qlib default setting and introduce the DropExtremeLabel (whose intuition is clearly stated in the paper and supplementary). We never exhaustively searched a preprocessing workflow, as it is not related to the research questions. 

(3) We use the same preprocessing for all the baselines.

In the future, I will not answer posts that are
- not related to our research questions or codes.
- Casually written.
- Too basic and too general. (pls discuss with your peers instead of posting to discuss with me.)
